// gatsby-cli/lib/reporter/reporters/ink/activities/activity.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _util = require("./util");

const generateActivityFinishedText = (name, activity) => {
  let successText = `${name} - ${(0, _util.calcElapsedTime)(activity.startTime)} s`;

  if (activity.status) {
    successText += ` — ${activity.status}`;
  }

  return successText;
};

var _default = reporter => name => {
  return {
    start: () => {
      reporter.setState(state => {
        return {
          activities: { ...state.activities,
            [name]: {
              type: `spinner`,
              status: ``,
              startTime: process.hrtime()
            }
          }
        };
      });
    },
    setStatus: status => {
      reporter.setState(state => {
        const activity = state.activities[name];
        return {
          activities: { ...state.activities,
            [name]: { ...activity,
              status: status
            }
          }
        };
      });
    },
    end: () => {
      const activity = reporter.state.activities[name];
      reporter.success(generateActivityFinishedText(name, activity));
      reporter.setState(state => {
        const activities = { ...state.activities
        };
        delete activities[name];
        return {
          activities
        };
      });
    }
  };
};

exports.default = _default;
